scalableminds/webknossos
# Async IO for HttpsDataVault, Fox Error Handling

- VaultPath.readBytes now returns Fox[Array[Byte]]
 - Adapted usages and implementations accordingly
 - Used the actually async http client `play ws` in the HttpsDataVault
 - 404 is now propagated as Fox.empty and fill-value is inserted there. Other errors are propagated as errors (leading to neutral gray rendered by the front-end rather than the fill value’s black). Note that this only works for the HttpsDataVault at this point

### TODO
 - [x] adapt vault path and data vault array interfaces
 - [x] adapt usages
    - [x] adapt opening arrays and reading headers
    - [x] adapt reading chunks
 - [x] adapt tests
 - [x] use non-blocking http request in HttpsDataVault
   - [x] pass ws in there
   - [x] use ws to get info and data
   - [x] basic auth
   - [x] what about connect timeout?
   - [x] ensure this works with gzip and brotli encodings
 - [x] 404 should be propagated as Fox.empty and use fill value where appropriate
 - [x] merge master into this after #7067 is in there

### URL of deployed dev instance (used for testing):
- https://vaultreadfox.webknossos.xyz

### Steps to test:
- Explore and view some remote datasets with different stores/data types, should still show data as before
- Self-streaming should now work (create zarr link in annotation, explore that link in the same webknossos instance)

### Issues:
- fixes #6906 
- fixes #6872
- fixes #7141

------
